backend/src/server.rs

```diff
@@ -0,0 +1,501 @@
+use std::collections::{HashMap};
+use std::cmp::{max, min};
+use std::io::Read;
+
+use iron::prelude::*;
+use iron::status;
+use router::Router;
+use persistent::State;
+use chrono::{Duration, NaiveDate, NaiveDateTime};
+use json::{self, JsonValue};
+
+
+use errors::*;
+use load::{Data, ServerData};
+use util::{start_idx, end_idx};
+
+const JS_DATE_FORMAT: &'static str = "%Y-%m-%dT%H:%M:%S.000Z";
+
+fn get_summary(_req_body: Option<JsonValue>, data: &ServerData) -> Result<JsonValue> {
+    let dates = data.summary_rustc.summary.iter()
+        .map(|s| {
+            s.date.format(JS_DATE_FORMAT).to_string()
+        })
+        .take(12)
+        .collect::<Vec<_>>();
+
+    // overall number for each week
+    let mut summaries = data.summary_benchmarks.summary.iter().enumerate().map(|(i, s)| {
+        let mut sum = 0.0;
+        let mut count = 0.0;
+        for (_, krate) in &s.by_crate {
+            if krate.contains_key("total") {
+                sum += krate["total"];
+                count += 1.0;
+            }
+        }
+
+        if data.summary_rustc.summary[i].by_crate["total"].contains_key("total") {
+            sum += 2.0 * data.summary_rustc.summary[i].by_crate["total"]["total"];
+            count += 2.0;
+        }
+
+        format!("{:.1}", sum / count)
+    }).collect::<Vec<_>>();
+
+    // per benchmark, per week
+    let mut breakdown = data.summary_benchmarks.summary.iter().enumerate().map(|(i, s)| {
+        let mut per_bench = HashMap::new();
+
+        for (crate_name, krate) in &s.by_crate {
+            let val = krate.get("total").map_or("0.0".to_string(), |total| format!("{:.1}", total));
+            per_bench.insert(crate_name.to_string(), JsonValue::from(val));
+        }
+
+        let bootstrap = if data.summary_rustc.summary[i].by_crate["total"].contains_key("total") {
+            format!("{:.1}", data.summary_rustc.summary[i].by_crate["total"]["total"])
+        } else {
+            "0.0".to_string()
+        };
+        per_bench.insert("bootstrap".to_string(), bootstrap.into());
+
+
+        per_bench.into()
+    }).collect::<Vec<JsonValue>>();
+
+    Ok(object! {
+        "total_summary" => summaries.remove(0),
+        "total_breakdown" => breakdown.remove(0),
```

Comment:
I'm a little worried that the order in summary_* isn't guaranteed. (Order in JS may have been always the same due to an implementation detail, but that may not be true here: or even if it is, we should not depend on it).